lib/utils/genmap/genmap.go

```diff
@@ -0,0 +1,317 @@
+	// is closed.
+	select {
+	case <-ent.init:
+		return ent.ptr.Load().read()
+	case <-m.ctx.Done():
+		return val, ErrGenMapClosed
+	case <-ctx.Done():
+		return val, ctx.Err()
+	}
+}
+
+// ensureEntry gets the entry for the target key if it exists, falling back to setting up a new one.
+func (m *GenMap[K, V]) ensureEntry(key K) (ent *entry[V], isNew bool) {
+	m.rw.RLock()
+	ent, ok := m.entries[key]
+	m.rw.RUnlock()
+	if ok {
+		return ent, false
+	}
+
+	// create an entry-level context that is a child of the
+	// main close context.
+	ctx, cancel := context.WithCancel(m.ctx)
+
+	newEnt := &entry[V]{
+		init:      make(chan struct{}),
+		wantRegen: make(chan struct{}, 1),
+		ctx:       ctx,
+		cancel:    cancel,
+	}
+
+	m.rw.Lock()
+	// check if another entry was set up concurrently
+	ent, ok = m.entries[key]
+	if !ok {
+		m.entries[key] = newEnt
+	}
+	m.rw.Unlock()
+	if ok {
+		cancel()
+		return ent, false
+	}
+
+	// start background opts for new entry
+	go m.generate(key, newEnt)
+
+	return newEnt, true
+}
+
+// generate is the background process that generates/regenerates the value associated with a given key.
+func (m *GenMap[K, V]) generate(key K, entry *entry[V]) {
+	defer func() {
+		m.rw.RLock()
+		current := m.entries[key]
+		m.rw.RUnlock()
+		if current != entry {
+			return
+		}
+
+		m.rw.Lock()
+		if m.entries[key] == entry {
+			delete(m.entries, key)
+		}
+		m.rw.Unlock()
+	}()
+
+	t := time.NewTimer(m.cfg.Jitter(m.cfg.RegenInterval))
+	defer t.Stop()
+
+	var init bool
+	var errs int
+	for {
+		// drain the reload channel since we're about to reload
+		select {
+		case <-entry.wantRegen:
+		default:
+		}
+
+		// reset the timer before loading to help compensate
+		// for slow generators.
+		t.Stop()
+		select {
+		case <-t.C:
+		default:
+		}
+		t.Reset(m.cfg.Jitter(m.cfg.RegenInterval))
+
+		// note that the generator is invoked with the map-level context, not the entry-level context.
+		// the entry-level context is only used to terminate background generators *between* regens as
+		// we want to avoid concurrent calls to Terminate generating spurious cancellation errors.
+		val, err := m.cfg.Generator(m.ctx, key)
+
+		// error values want to be reloaded, but reloading in a hot loop on errors is
+		// potentially problematic. as a compromise, we store a signal channel alongside
+		// error values that gets triggered by readers. this allows us to attempt regen
+		// relatively promptly for those values that are in high/frequent demand, while
+		// also preventing dangling/low-demand values from generating excess load due
+		// to constant regeneration.
+		var regenOnErr chan struct{}
+		if err != nil {
+			errs++
+			regenOnErr = make(chan struct{}, 1)
+		} else {
+			errs = 0
+		}
+
+		// update the atomic pointer to store the newly generated value.
+		entry.ptr.Store(&value[V]{
+			val:        val,
+			err:        err,
+			regenOnErr: regenOnErr,
```

Comment:
Can't we just use `entry.wantRegen` in lieu of a dedicated `regenOnErr` channel? Every time we call `(*value[_]).read` we have the entry available.

Reply:
`regenOnErr` is about the value observed by the signaler.  `wantRegen` is about the external state that the key maps to.  `regenOnErr` is saying "this specific value should be regenerated".  `wantRegen` is saying, "I think there is a newer value available for this key".  So for `regenOnErr` we only care about the signal if the value being observed by the sender is the same value that is still being advertised by the map.  On the other hand, `wantRegen` demands that we generate a value with a proper "happens after" relationship to the signaller sending their signal.

Reply:
Is there a practical difference between the two tho? `wantRegen` is drained before generating the new value, `regenOnErr` is replaced right afterwards, and they both trigger regeneration in the same way.